src/lib.rs

```diff
@@ -848,10 +860,117 @@ impl LineSpans for alloc::string::String {
     }
 }
 
+#[derive(Debug, Clone)]
+/// Owning pointer to a string that provides O(1) access to line slices
+///
+/// ```
+/// use line_span::CachedLines;
+/// let cache = CachedLines::without_ending(String::from("hello\nworld"));
+/// assert_eq!(&cache[0], "hello");
+/// assert_eq!(&cache[1], "world");
+/// ```
+///
+/// Slices might or might not include line breaks depending on a function used to construct it
+///
+/// - [`without_ending`](CachedLines::without_ending)
+/// - [`with_ending`](CachedLines::with_ending)
+///
+///
+#[cfg(feature = "alloc")]
+pub struct CachedLines {
+    content: alloc::string::String,
+    splits: alloc::vec::Vec<Range<usize>>,
+}
+
+#[cfg(feature = "alloc")]
+impl CachedLines {
+    /// Splits a [`String`](alloc::string::String) into lines, slices will include linebreaks
+    ///
+    /// ```
+    /// use line_span::CachedLines;
+    /// let cache = CachedLines::with_ending(String::from("hello\nworld"));
+    /// assert_eq!(&cache[0], "hello\n");
+    /// assert_eq!(&cache[1], "world");
+    /// ```
+    pub fn with_ending(content: alloc::string::String) -> Self {
+        let splits = content
+            .line_spans()
+            .map(|s| s.range_with_ending())
+            .collect::<alloc::vec::Vec<_>>();
+        Self { splits, content }
+    }
+
+    /// Splits a [`String`](alloc::string::String) into lines, slices will not include linebreaks
+    ///
+    /// ```
+    /// use line_span::CachedLines;
+    /// let cache = CachedLines::without_ending(String::from("hello\nworld"));
+    /// assert_eq!(&cache[0], "hello");
+    /// assert_eq!(&cache[1], "world");
+    /// ```
+    pub fn without_ending(content: alloc::string::String) -> Self {
+        let splits = content
+            .line_spans()
+            .map(|s| s.range())
+            .collect::<alloc::vec::Vec<_>>();
+        Self { splits, content }
+    }
+
+    /// Returns a reference to a line inside a contained string
+    ///
+    /// Returns [`None`] if index is out of bounds
+    /// ```
+    /// use line_span::CachedLines;
+    /// let cache = CachedLines::with_ending(String::from("hello\nworld"));
+    /// assert_eq!(cache.get(0), Some("hello\n"));
+    /// assert_eq!(cache.get(3), None);
+    /// ```
+    pub fn get(&self, index: usize) -> Option<&str> {
+        self.content.get(self.splits.get(index)?.clone())
```

Comment:
(in extension of the previous comment)

Then have this be something like:

```suggestion
    pub fn get(&self, index: usize) -> Option<LineSpan<'_>> {
        // i.e. from splits get LineSpanWithoutText + self.contents => LineSpan
        ...
```

Reply:
Yea, for `get` this works.